text/0000-compile-time-asserts.md

```diff
@@ -0,0 +1,115 @@
+- Feature Name: compile_time_asserts
+- Start Date: 2015-07-30
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+If the compiler can detect at compile-time that something will always
+cause a `debug_assert` or an `assert` it should instead
+insert an unconditional runtime-panic and issue a warning.
+
+# Definition of constant evaluation context
+
+There are exactly three places where an expression needs to be constant.
+
+- the initializer of a constant `const foo: ty = EXPR` or `static foo: ty = EXPR`
+- the size of an array `[T; EXPR]`
+- the length of a repeat expression `[VAL; LEN_EXPR]`
```

Comment:
This isn't a complete list... off the top of my head, patterns and enum discriminants are also required to be constant.